### PR TITLE
feat(appmesh): remove from*Name() methods and replace with from*Attributes()

### PR DESCRIPTION
**Pull request by dfezzie**
Follow Up from https://github.com/aws/aws-cdk/pull/10879

BREAKING CHANGE: all `fromResourceName()` methods in the AppMesh module have been replaced with `fromResourceAttributes()`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts [hunk 1]
```diff
@@ -151,18 +148,26 @@ class ImportedGatewayRoute extends cdk.Resource implements IGatewayRoute {
   public gatewayRouteArn: string;
 
   /**
-   * The VirtualGateway the GatewayRoute belongs to
+   * The VirtualGateway the GatewayRoute is associated with
    */
   public virtualGateway: IVirtualGateway;
 
-  constructor(scope: Construct, id: string, props: GatewayRouteAttributes) {
+  constructor(scope: Construct, id: string, attrs?: GatewayRouteAttributes, gatewayRouteArn?: string) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:155]: change to props?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts [hunk 2]
```diff
@@ -114,26 +121,16 @@ export class GatewayRoute extends cdk.Resource implements IGatewayRoute {
...
-   * The name of the mesh this GatewayRoute is associated with
+   * The name of the VirtualGateway this GatewayRoute is associated with
    */
-  readonly meshName?: string;
-
-  /**
-   * The name of the Virtual Gateway this GatewayRoute is associated with
-   */
-  readonly virtualGateway?: IVirtualGateway;
+  readonly virtualGateway: IVirtualGateway;
```
Comment (eladb) [inline packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:133]: In EKS, for example we actually only take ARNs here and not strongly-typed objects.

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts (hunk 1 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:155]: I can do that. I changed it since the `fromAttributes` methods asked for the parameter to be named `attrs`

Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:155]: Is there a general rule of thumb when it should be `props` vs `attrs`?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts (hunk 2 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:133]: @skinny85 asked for these to be strongly typed. An ARN would work for this, but this using the constructs is more explicit.

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:155]: I think what Elad is saying is to add an `ImportedGatewayRouteProps` interface that extends `GatewayRouteAttributes` and contains the `gatewayRouteArn` property, instead of having 4 arguments to this constructor.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts [hunk 3]
```diff
@@ -26,14 +26,9 @@ export interface IVirtualGateway extends cdk.IResource {
...
   /**
-   * The mesh which the VirtualGateway belongs to
+   * The Mesh which the VirtualGateway belongs to
    */
   readonly mesh: IMesh;
-
-  /**
-   * Utility method to add a new GatewayRoute to the VirtualGateway
-   */
-  addGatewayRoute(id: string, route: GatewayRouteBaseProps): GatewayRoute;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:36]: Why was this removed?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts (hunk 1 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:155]: Got it. Makes sense

File: packages/@aws-cdk/aws-appmesh/test/test.gateway-route.ts [hunk 4]
```diff
@@ -133,20 +133,36 @@ export = {
     },
   },
 
-  'Can export and import GatewayRoutes and perform actions'(test: Test) {
+  'Can import Gateway Routes using ARN and attributes'(test: Test) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.gateway-route.ts:136]: Can you leave this test unchanged, and add a new one for `fromGatewayRouteAttributes` please?

File: packages/@aws-cdk/aws-appmesh/test/test.route.ts [hunk 5]
```diff
@@ -4,55 +4,33 @@ import { Test } from 'nodeunit';
...
-      mesh,
+    const stack = new cdk.Stack(app, 'Imports', {
+      env: { account: '123456789012', region: 'us-east-1' },
     });
+    const meshName = 'test-mesh';
+    const virtualRouterName = 'test-virtual-router';
+    const routeName = 'test-route';
+    const arn = `arn:aws:appmesh:us-east-1:123456789012:mesh/${meshName}/virtualRouter/${virtualRouterName}/gatewayRoute/${routeName}`;
 
-    const service1 = new appmesh.VirtualService(stack, 'service-1', {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.route.ts:20]: Same comment here - leave the old test as-is, and create a new one for `fromRouteAttributes`.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts [hunk 6]
```diff
@@ -239,20 +239,31 @@ export = {
...
+    test.equal(virtualGateway1.virtualGatewayName, virtualGatewayName);
 
+    // WHEN
+    const virtualGateway2 = appmesh.VirtualGateway.fromVirtualGatewayArn(
+      stack, 'importedGateway2', arn);
     // THEN
-    test.equal(virtualGateway2.mesh.meshName, 'testMesh');
-    test.equal(virtualGateway2.virtualGatewayName, 'test-gateway');
+    test.equal(virtualGateway2.mesh.meshName, meshName);
+    test.equal(virtualGateway2.virtualGatewayName, virtualGatewayName);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:266]: Same comment here - leave the old test as-is, and create a new one for `fromVirtualGatewayAttributes`.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts [hunk 7]
```diff
@@ -102,55 +102,29 @@ export = {
...
+      mesh: appmesh.Mesh.fromMeshName(stack, 'Mesh', meshName),
+      virtualNodeName: virtualNodeName,
     });
+    // THEN
+    test.equal(virtualNode1.mesh.meshName, meshName);
+    test.equal(virtualNode1.virtualNodeName, virtualNodeName);
 
+    // WHEN
+    const virtualNode2 = appmesh.VirtualNode.fromVirtualNodeArn(
+      stack, 'importedVirtualNode2', arn);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:124]: Same comment here - leave the old test as-is, and create a new one for `fromVirtualNodeAttributes `.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-router.ts [hunk 8]
```diff
@@ -304,15 +304,29 @@ export = {
     },
   },
 
-  'can import a virtual router'(test: Test) {
+  'Can import Virtual Routers using ARN and attributes'(test: Test) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-router.ts:307]: Same comment here - leave the old test as-is, and create a new one for `fromVirtualRouterAttributes `.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts [hunk 9]
```diff
@@ -1,38 +1,31 @@
...
 
-    const stack2 = new cdk.Stack();
-    appmesh.VirtualService.fromVirtualServiceName(stack2, 'imported-virtual-service', mesh.meshName, service.virtualServiceName);
-
-    // Nothing to do with imported virtual service yet
+    // WHEN
+    const virtualRouter2 = appmesh.VirtualRouter.fromVirtualRouterArn(stack, 'importedVirtualRouter2', arn);
+    // THEN
+    test.equal(virtualRouter2.mesh.meshName, meshName);
+    test.equal(virtualRouter2.virtualRouterName, virtualServiceName);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:28]: Same comment here - leave the old test as-is, and create a new one for `fromVirtualServiceAttributes`.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Good first attempt, but I'd like us to use a different structure that doesn't have these weird `if`s inside of the imported classes.

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:133]: Yes, we always use the strongly-typed properties for these, surprised EKS does something different.

File: packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts (hunk 7 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:124]: The old test here does not make much sense. It adds a listener to `node2` but never does any sort of validation on it. Adding the listener to the imported class does not actually make much sense, as it is modeled outside of the CDK and making a mutating call on it would not be valid.

I can revert it, but it does not provide much value.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts (hunk 3 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:36]: My original thought was that this does not make sense to have in the `IVirtualGateway` resource. If we were to implement the inline class you suggested, we would need to also implement the `addGatewayRoute` method for that class, right? 

Thinking about it more, calling `addGatewayRoute` would be a valid call for imported virtual gateways, but you could accomplish the same thing by passing the imported gateway to the Gateway Route constructor. What are your thoughts on this?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts [hunk 10]
```diff
@@ -151,18 +148,26 @@ class ImportedGatewayRoute extends cdk.Resource implements IGatewayRoute {
...
   public virtualGateway: IVirtualGateway;
 
-  constructor(scope: Construct, id: string, props: GatewayRouteAttributes) {
+  constructor(scope: Construct, id: string, attrs?: GatewayRouteAttributes, gatewayRouteArn?: string) {
     super(scope, id);
-    if (props.gatewayRouteArn) {
-      this.gatewayRouteArn = props.gatewayRouteArn;
-      this.gatewayRouteName = cdk.Fn.select(4, cdk.Fn.split('/', cdk.Stack.of(scope).parseArn(props.gatewayRouteArn).resourceName!));
-      this.virtualGateway = VirtualGateway.fromVirtualGatewayArn(this, 'virtualGateway', props.gatewayRouteArn);
+    if (gatewayRouteArn) {
```
[the remark this answers is not in the document]
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:157]: Got it. I'll make this change

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:36]: The way we deal with this usually in the CDK is to have an abstract module-private class called `<Resource>Base` (so in this case, `VirtualGatewayBase`) that implements the resource interface (so `IVirtualGateway` in this case), and that class contains the implementation of the methods like `addGatewayRoute`. In that case, the inline class should extend from `VirtualGatewayBase` instead of `cdk.Resource`:

```ts
  public static fromVirtualGatewayAttributes(scope: Construct, id: string, attrs: VirtualGatewayAttributes): IVirtualGateway {
    return new class extends VirtualGatewayBase {
      // ...
```

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:36]: Apparently there's already `VirtualGatewayBase`, so your work is easier: https://github.com/aws/aws-cdk/blob/8c9321b833c1dc9eed1c2e33fe3715cb0b181630/packages/%40aws-cdk/aws-appmesh/lib/virtual-gateway.ts#L75

Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:36]: Looking at other classes, it looks like they implement a `ResourceBase` class in instances like this. I will do the same here

File: packages/@aws-cdk/aws-appmesh/test/test.route.ts (hunk 5 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/test/test.route.ts:20]: Since these tests are not actually asserting anything, I'm going to opt to not add them back. I did split the new test into two separate ones that test import by ARN and import by attributes separately. Let me know if you disagree with this path.

**Review by skinny85 — APPROVED** [review]
Looks great, thanks @dfezzie !

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 11]
```diff
@@ -307,3 +307,21 @@ gateway.addGatewayRoute('gateway-route-grpc', {
   }),
 });
 ```
+
+## Importing Resources
+
+Each resource comes with two static methods for importing a reference to an existing App Mesh resource.
+There are two static methods, `from<Resource>Arn` and `from<Resource>Attributes` where the `<Resource>` is replaced with the resource name.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:314]: I guess `Mesh` is an exception to the rule...?

File: packages/@aws-cdk/aws-appmesh/lib/route.ts [hunk 12]
```diff
@@ -111,21 +116,37 @@ export interface RouteProps extends RouteBaseProps {
...
-    return new ImportedRoute(scope, id, { routeArn });
+    return new class extends cdk.Resource implements IRoute {
+      readonly routeArn = routeArn;
+      readonly mesh = Mesh.fromMeshName(this, 'Mesh', cdk.Fn.select(0, cdk.Fn.split('/', cdk.Stack.of(scope).parseArn(routeArn).resourceName!)));
+      readonly virtualRouter = VirtualRouter.fromVirtualRouterAttributes(this, 'VirtualRouter', {
+        mesh: this.mesh,
+        virtualRouterName: cdk.Fn.select(2, cdk.Fn.split('/', cdk.Stack.of(scope).parseArn(routeArn).resourceName!)),
+      });
+      readonly routeName = cdk.Fn.select(4, cdk.Fn.split('/', cdk.Stack.of(scope).parseArn(routeArn).resourceName!));
+    }(scope, id);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route.ts:130]: The expression `cdk.Fn.split('/', cdk.Stack.of(scope).parseArn(routeArn).resourceName!)` is repeated like 5 times here... can you extract it to a local variable instead?

**Review by skinny85 — COMMENTED** [review]
Snuck in 2 minor last comments since Mergify dismissed my review for some reason 😜

**Comment by dfezzie** [comment]
> Snuck in 2 minor last comments since Mergify dismissed my review for some reason 😜

It was due to the README not being updated with the `feat(...)` tag. Figure it would be good to get those documented. 😁

File: packages/@aws-cdk/aws-appmesh/README.md (hunk 11 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/README.md:314]: hm. Fair point..

File: packages/@aws-cdk/aws-appmesh/lib/route.ts (hunk 12 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route.ts:130]: Yeah, would be more readable for sure, good call

File: packages/@aws-cdk/aws-appmesh/README.md (hunk 11 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/README.md]: Updated to include `Mesh`

**Review by skinny85 — DISMISSED** [review]

**Review by skinny85 — APPROVED** [review]
